server/src/main/resources/rules/rules.js

```diff
@@ -2447,44 +2455,62 @@ var Autobind = {
         var best = null;
         var total_poolquantity = Number.MAX_VALUE;
         var best_avg_prio = 0;
+        var best_num_host_specific = 0;
+        var best_num_virt_only = 0;
+        var virt_only_found = false;
+        var host_specific_found = false;
+        var new_best_found = false;
 
         for (var i = 0; i < all_groups.length; i++) {
             var group = all_groups[i];
             var group_avg_prio = group.get_average_priority();
             var intersection = this.get_common_products(installed, group).length;
             var group_poolquantity = group.get_total_quantity();
-            // Choose group that provides the most installed products
-            if (intersection > max_provide) {
+            var group_num_host_specific = group.num_host_specific();
+            var group_num_virt_only = group.num_virt_only();
+            if (intersection <= 0 ||
+                (host_specific_found && group_num_host_specific < best_num_host_specific) ||
+                (virt_only_found && group_num_virt_only < best_num_virt_only)) {
+                // Skip this group if we've found virt or host_specific and this group is not.
+                continue;
+            }
+
+            new_best_found = false;
+            if (group_num_host_specific > best_num_host_specific) {
+                host_specific_found = true;
+                new_best_found = true;
+            } else if (group_num_host_specific < best_num_host_specific) {
```

Comment:
Seems the javascript files sneak around checkstyle. All of these "else if..." bits should be on a new line, as such:

```
if (/*something*/) {
    // Do something
}
else if (/* something else */) {
    // Do the other thing
}
else if (...) {
   ...
```